Optimise common operations on Array and List

adriaanm opened the pull request:
Consolidates the following PRs:
- List.mapConserve avoids ListBuffer creation (#5650)
- Array slice optimisations (#5652)
- Optimised implementation of List.filter/filterNot (#5653)

rorygraves committed:
Avoid creating ListBuffer in List.mapConserve

Co-Authored-By: Jason Zaugg <jzaugg@gmail.com>

rorygraves committed:
Optimised implementation of List.filter/filterNot

mkeskells committed:
Test IndexedSeq, including ArrayOps, WrappedArray.

Specifically, the `slice` and `take` methods.

mkeskells committed:
Optimize slice and take in ArrayOps, WrappedArray

use Array block copy operations rather than builder/iterator

adriaanm commented:
I suggest we either close the other PRs, merge this one, and add the benchmarks, or incorporate these commits into one of the other PRs and close this one.

rorygraves commented:
@adriaanm  The List.filter/filterNot benchmark already exists on the https://github.com/scala/scala/pull/5653 PR ListBenchmark.scala

rorygraves committed:
Add benchmark for List.mapConserve

adriaanm commented:
Ok, I wasn't clear whether that was the last word on that. I cherry-picked the commits, so then the benchmark is in here too.

adriaanm commented:
Mima should be happy with these new commits. I haven't looked into whether each whitelist item is good to go. Leaving that up to @szeiger

rorygraves commented:
Looks good.

adriaanm commented:
Thanks, sounds like I can safely close the other PRs, since all their commits have been included here? Will get this one merged asap.

rorygraves commented:
Yes they can all be safely closed. 
:)


> On 2 Feb 2017, at 18:33, Adriaan Moors <notifications@github.com> wrote:
> 
> Thanks, sounds like I can safely close the other PRs, since all their commits have been included here? Will get this one merged asap.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

adriaanm commented:
@szeiger, are you ready to sign off on this? /cc @retronym

szeiger commented on bincompat-forward.whitelist.conf:
```diff
@@ -628,6 +628,319 @@ filter {
+    },
+    {
+        matchName="scala.collection.mutable.WrappedArray.sliceImpl"
+        problemName=DirectMissingMethodProblem
```
This is a breaking change. WrappedArrray is public and the new method is protected. As discussed before, all these changes need to be moved down into a private implementation class.

szeiger commented on bincompat-forward.whitelist.conf:
```diff
@@ -628,6 +628,319 @@ filter {
+    },
+    {
+        matchName="scala.collection.mutable.WrappedArray.emptyImpl"
+        problemName=DirectMissingMethodProblem
```
Same here.

szeiger commented on bincompat-forward.whitelist.conf:
```diff
@@ -628,6 +628,319 @@ filter {
+    },
+    {
+        matchName="scala.collection.mutable.WrappedArray.slice"
+        problemName=IncompatibleResultTypeProblem
```
This is also a problem.

szeiger commented:
I'll try to fix these remaining issues after lunch. Should be ready in a few hours.

szeiger committed:
Handle WrappedArray the same way as ArrayOps for binary compatibility

We introduce a package-private superclass with the overridden
implementations. This is public at the bytecode level, so it needs
to be whitelisted.

szeiger commented:
https://github.com/adriaanm/scala/pull/26

adriaanm commented:
@szeiger I merged your own fixes for your review -- thanks!

adriaanm commented:
Looks like this is causing LinkageErrors in user code.  Should we revert, declare 2.11.9 a dud and cut 2.11.10?

# Detailed diagnosis by @xuwei-k:
https://github.com/xuwei-k/Scala-2.11.9-FilteredTraversableInternal

# Explanation by @lrytz
The reason why a reference to `FilteredTraversableInternal` appears in @xuwei-k's example:
```scala
List(1).map(identity).filter(_ => true)
```
In erasure type checking, `List(1).map(identity)` has type `Object` (the result type is `That` which is a type parameter), so selecting the `filter` method requires a cast. This is done by [erasure's `adaptMember`](https://github.com/scala/scala/blob/v2.11.9/src/compiler/scala/tools/nsc/transform/Erasure.scala#L674-L676), which casts to the filter method symbol's `owner`, which is now `FilteredTraversableInternal`.

szeiger commented:
@lrytz So we now have a situation that looks binary compatible in bytecode but leads to incompatibilities due to additional information in ScalaSig?

lrytz commented:
oh, or maybe it's a NoClassDefFoundErro, which is a linkage error, i'm on
the phone :)


On Mon, 3 Apr 2017 at 17:02 Lukas Rytz <lukas.rytz@lightbend.com> wrote:

> it's not in the scala signature, but bytecode instructions (cast, call
> receiver type). i'm not sure it's a binary incompatibility by our defnition
> (no linkage error), as you get a ClassNotFound. but the end result is an
> incompatibility we don't want to have in a minor release.
>
> On Mon, 3 Apr 2017 at 16:51 Stefan Zeiger <notifications@github.com>
> wrote:
>
>> @lrytz <https://github.com/lrytz> So we now have a situation that looks
>> binary compatible in bytecode but leads to incompatibilities due to
>> additional information in ScalaSig?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scala/scala/pull/5664#issuecomment-291167072>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AAHTVMdhEfsrDgq9snlldbMGqHMx1KPIks5rsQd_gaJpZM4Lwl1L>
>> .
>>
>

szeiger commented:
"in the bytecode instructions" is the effect, not the cause. Binary compatibility only matters when you have separate compilation, in which case the compiler must be getting the idea of casting to `FilteredTraversableInternal` from somewhere in the class file (ScalaSig?). If I understand you correctly the erased return type of `filter` is `Object` but somehow the compiler knows that the return type is really a type parameter that was instantiated as `FilteredTraversableInternal` and inserts a cast to that type. This implies that looking at the erased types is not sufficient for checking binary compatibility in MiMa. OTOH, why isn't there a bridge method for `filter` if it's overridden in `FilteredTraversableInternal`? Shouldn't we see two copies of this method (one returning `Object` and the other one returning `FilteredTraversableInternal` in all subclasses? Or maybe we do but MiMa doesn't recognize it correctly? Or is it missing because they type is private and therefore must not leak (but still does due to a compiler bug)?

(I'm only speculating here. I'd need more time to look into this but I don't expect to find any until after ScalaDays. For now the only safe solution appears to be never to add a new superclass, even if it's declared private and not supposed to leak.)

lrytz commented:
> If I understand you correctly the erased return type of filter is Object but somehow the compiler knows that the return type is really a type parameter that was instantiated as FilteredTraversableInternal and inserts a cast to that type

My explanation was unclear, sorry: the return type of `map` is its `That` type parameter. When type-checking `List(1).map(identity).filter` during erasure, the qualifier `List(1).map(identity)` gets type `Object`, so the selection `qual.filter` needs a cast: `qual.asInstanceOf[<owner-of-filter>].filter`. This happens here: https://github.com/scala/scala/blob/v2.11.9/src/compiler/scala/tools/nsc/transform/Erasure.scala#L674-L676

> This implies that looking at the erased types is not sufficient for checking binary compatibility in MiMa

I would say that MiMa did its job correctly: it identified the class being added, it was whitelisted. Any reference to a new class is a binary incompatibility.

> Shouldn't we see two copies of this method (one returning Object and the other one returning FilteredTraversableInternal) in all subclasses?

The return type of `FilteredTraversableInternal.filter` in bytecode is `Object`.

szeiger commented:
Thanks, that makes it clearer. So we'd have to change this to cast to the actual type of the qualifier instead of the method's owner (which can be a supertype) to prevent these leaks?

> I would say that MiMa did its job correctly: it identified the class being added, it was whitelisted. Any reference to a new class is a binary incompatibility.

I was hoping we could change MiMa to ignore new non-accessible superclasses and treat this case as binary compatible but with these owner casts leaking superclasses the only safe situation would be one where all methods implemented by the new superclass are overridden in all subclasses.

lrytz commented:
> Thanks, that makes it clearer. So we'd have to change this to cast to the actual type of the qualifier instead of the method's owner (which can be a supertype) to prevent these leaks?

This cannot easily be done, unfortunately. At the beginning of erasure, all types of trees are set to `null`, and then the type checker is run again, this time assigning only erased types. So `List(1).map(identity)` really gets type `Object`, I don't think there's a way to get `List` at this point.

We could remember the type that the tree had after typer (`List[Int]`) and use an erased version of that for the cast, but that sounds like a hack. Incidentally, we need to do exactly that in order to avoid `IllegalAccessError`s for selections of Java-defined methods: if the method is defined in a package-protected class, it might not be accessible. I very recently worked on this: https://github.com/scala/scala/pull/5792. I'm not sure we should use this hack in general, for Scala-only code.

szeiger commented:
After further consideration 

> the only safe situation would be one where all methods implemented by the new superclass are overridden in all subclasses

even this wouldn't be safe because it's not transitive. Adding a new superclass with overrides of all methods in all subclasses would be considered safe, adding a new override to an existing private class is safe, but the combination of both breaks binary compatibility.
